### PR TITLE
add rolling options to rolling_apply

### LINKED ISSUE #1938: Add missing arguments to rolling other functions
Right now, (as far as I can tell) all the rolling transforms have `weights`, `min_periods`, and `center` as arguments, except for the following:
- `rolling_median`
- `rolling_quantile`
- `rolling_skew`
- `rolling_apply`

It would be great if those arguments also get included there.

### PR DESCRIPTION
**Pull request by marcvanheerden**
Still to do:

- multiply in the weights 
- add more tests

fixes #1938

### CODE REVIEW

**Comment by marcvanheerden** [comment]
Should be ready for review by tomorrow morning

**Comment by marcvanheerden** [comment]
Sorry @ritchie46  taking me a little longer. What do you want to do with the types and weights? 
I'm thinking this:
if no weights -> T
if weights and T is f32 -> f32
if weights and T is anything else -> f64

Let me know what you think?

**Comment by ritchie46** [comment]
Seems good to me @marcvanheerden :+1:

**Comment by marcvanheerden** [comment]
Should be ready for review, I'm just going to run some comparisons to see if there is a loss in performance, should be done in the next couple of hours

File: polars/polars-core/src/chunked_array/ops/rolling_window.rs [hunk 1]
```diff
@@ -273,45 +277,125 @@ mod inner_mod {
...
+
+                        // Safety.
+                        // ptr is not dropped as we are in scope
+                        // We are also the only owner of the contents of the Arc
+                        // we do this to reduce heap allocs.
+                        unsafe {
+                            *ptr = arr_window;
+                        }
+
+                        let s = if size == options.window_size {
```
Comment (ritchie46) [inline polars/polars-core/src/chunked_array/ops/rolling_window.rs:348]: This is most of the cases right?

File: polars/polars-core/src/chunked_array/ops/rolling_window.rs [hunk 2]
```diff
@@ -273,45 +277,125 @@ mod inner_mod {
...
+                                    .f32()
+                                    .unwrap()
+                                    .into_iter()
+                                    .take(series_container.len())
+                                    .collect(),
+                            };
+                            f(&series_container.multiply(&weights_cutoff).unwrap())
+                        };
+
+                        let out = self.unpack_series_matching_type(&s)?;
```
Comment (ritchie46) [inline polars/polars-core/src/chunked_array/ops/rolling_window.rs:368]: You found how to get the generic type. ;)

**Review by ritchie46 — COMMENTED** [review]
Thank @marcvanheerden I think it looks good.

This will likely not be most performant once we supply weights because of dynamic dispatch, but its at least consistent now with the other `rolling_`, thank you!

File: polars/polars-arrow/src/kernels/rolling/no_nulls.rs [hunk 3]
```diff
@@ -163,6 +163,10 @@ pub(crate) fn compute_var<T>(vals: &[T]) -> T
 where
     T: Float + std::iter::Sum,
 {
+    dbg!(vals
```
[the remark this answers is not in the document]
Comment (marcvanheerden) [reply polars/polars-arrow/src/kernels/rolling/no_nulls.rs:166]: thanks!

**Comment by marcvanheerden** [comment]
I thought the performance would be poor but now I'm not so sure. 

I tested it with this:
```
import polars as pl
import time 

s = pl.Series("A", [x for x in range(100000)], dtype = pl.Int32)

start = time.time()
out = s.rolling_apply(window_size=3, function=lambda s: s.std())
print("rolling apply std")
print(time.time() - start)

start = time.time()
out = s.rolling_apply(window_size=3,
        function=lambda s: s.std(),
        weights = [1.0, 2.0, 3.0]
        )
print("rolling apply weights")
print(time.time() - start)
```

For the first piece, I get ~4.7s on the master branch and this branch. 
The second piece only runs on this branch and is slightly less than 4.7s.

Do you think the inefficiencies are getting optimised away or is this not a good example?

File: polars/polars-core/src/chunked_array/ops/rolling_window.rs (hunk 1 above)
Comment (marcvanheerden) [reply polars/polars-core/src/chunked_array/ops/rolling_window.rs:348]: Yes, should I try rewrite to avoid doing this check in every iteration?

**Comment by ritchie46** [comment]
> I thought the performance would be poor but now I'm not so sure.
> 
> I tested it with this:
> 
> ```
> import polars as pl
> import time 
> 
> s = pl.Series("A", [x for x in range(100000)], dtype = pl.Int32)
> 
> start = time.time()
> out = s.rolling_apply(window_size=3, function=lambda s: s.std())
> print("rolling apply std")
> print(time.time() - start)
> 
> start = time.time()
> out = s.rolling_apply(window_size=3,
>         function=lambda s: s.std(),
>         weights = [1.0, 2.0, 3.0]
>         )
> print("rolling apply weights")
> print(time.time() - start)
> ```
> 
> For the first piece, I get ~4.7s on the master branch and this branch. The second piece only runs on this branch and is slightly less than 4.7s.
> 
> Do you think the inefficiencies are getting optimised away or is this not a good example?

I think it would be intersting to compate with rolling_std


5 seconds is slow

**Comment by marcvanheerden** [comment]
Sure, it's fractions of a second, have to rebuild so will confirm soon. 5s is super slow, I just meant that this PR doesn't appear to have made it worse.

**Comment by ritchie46** [comment]
> Sure, it's 0.002s. 5s is super slow, I just meant that this PR doesn't appear to have made it worse. Do you want me to try some optimisation in this PR or separately?

Haha, no thank you. I don't think we can do much here. Slow it is :) Thanks for the PR!

**Comment by marcvanheerden** [comment]
the checks that failed seem to be unrelated to the code:
```
error: could not download file from 'https://static.rust-lang.org/rustup/release-stable.toml' 
```
Can you kick it off again or do I need to push a dummy commit to force it?

**Pull request merged**